File: sourcerer-io/src/main/java/sourcerer/io/Value64.java
/*
 * Copyright 2016 Layne Mobile, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sourcerer.io;

import com.google.common.base.Objects;

import java.io.IOException;
import java.util.List;

import okio.ByteString;

public final class Value64 implements Writeable {
    private static final Reader.Parser<Value64> PARSER = new Reader.Parser<Value64>() {
        @Override public final Value64 parse(Reader reader) throws IOException {
            String base64 = reader.readString();
            String value = ByteString.decodeBase64(base64)
                    .utf8();
            return new Value64(value, base64);
        }
    };

    private final String value;
    private final String base64;

    private Value64(String value) {
        this(value, ByteString.encodeUtf8(value).base64());
    }

    private Value64(String value, String base64) {
        this.value = value;
        this.base64 = base64;
    }

    public static Value64 from(int value) {
        return new Value64(Integer.toString(value));
    }

    public static Value64 from(String value) {
        return new Value64(value);
    }

    public static Value64 read(Reader reader) throws IOException {
        return PARSER.parse(reader);
    }

    public static List<Value64> readList(Reader reader) throws IOException {
        return reader.readList(PARSER);
    }

    public String value() {
        return value;
    }

    public int intValue() throws NumberFormatException {
        return Integer.parseInt(value);
    }

    @Override public void writeTo(Writer writer) throws IOException {
        writer.writeString(base64);
    }

    @Override public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Value64)) return false;
        Value64 value64 = (Value64) o;
        return Objects.equal(value, value64.value);
    }

    @Override public int hashCode() {
        return Objects.hashCode(value);
    }
}

File: sourcerer/io/src/main/java/sourcerer/io/Value64.kt
/*
 * Copyright (C) 2018 evove.tech
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package sourcerer.io

import com.google.common.base.Objects
import okio.ByteString
import sourcerer.io.Reader.Companion.parser
import java.io.IOException

class Value64
private constructor(
    private val value: String,
    private val base64: String = ByteString.encodeUtf8(value).base64()
) : Writeable {

    fun value(): String {
        return value
    }

    @Throws(NumberFormatException::class)
    fun intValue(): Int {
        return Integer.parseInt(value)
    }

    @Throws(IOException::class)
    override fun writeTo(writer: Writer) {
        writer.writeString(base64)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Value64) return false
        val value64 = other as Value64?
        return Objects.equal(value, value64!!.value)
    }

    override fun hashCode(): Int {
        return Objects.hashCode(value)
    }

    companion object {
        private val PARSER = parser { reader ->
            val base64 = reader.readString()
            val value = ByteString.decodeBase64(base64)!!
                    .utf8()
            Value64(value, base64)
        }

        fun from(value: Int): Value64 {
            return Value64(Integer.toString(value))
        }

        fun from(value: String): Value64 {
            return Value64(value)
        }

        @Throws(IOException::class)
        fun read(reader: Reader): Value64? {
            return PARSER.parse(reader)
        }

        @Throws(IOException::class)
        fun readList(reader: Reader): List<Value64> {
            return reader.readList(PARSER)
        }
    }
}
